File: SearchInRotatedSortedArray.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, r = 0, len(nums)-1
        while l <= r:
            m = (l+r)//2
            if nums[m] == target:
                return m
            elif nums[m] > target:
                if nums[m] < nums[r]:
                    r = m - 1
                else:
                    if nums[r] < target:
                        r = m - 1
                    else:
                        l = m + 1
            else:
                # m == l
                if nums[m] >= nums[l]:
                    l = m + 1
                else:
                    if nums[l] > target:
                        l = m + 1
                    else:
                        r = m - 1
        return -1

nums = [1, 3]
target = 3
print Solution().search(nums, target)

File: cpp/Search in Rotated Sorted Array.cpp
using namespace std;

class Solution {
public:
	int search(int A[], int n, int target) {
		int first = 0, last = n;
		while (first != last) {
			const int mid = (first + last) >> 1;
			if (A[mid] == target)
				return mid;
			if (A[first] <= A[mid]) {
				if (A[first] <= target && target < A[mid])
					last = mid;
				else
					first = mid + 1;
			} else {
				if (A[mid] < target && target <= A[last - 1])
					first = mid + 1;
				else 
					last = mid;
			}
		}
		return - 1;
	}
}; 